### ISSUE TITLE
Cursor in wrong location after inserting `Tabnine` completion items

### ISSUE THREAD

**Issue by mystilleef** [tzachar/cmp-tabnine#15]
For  some time now, I've noticed the cursor doesn't move to the end of a `tabnine` completion insertion. This happens especially when the completion ends with a closing bracket  or parenthesis like `)` or `}`. I've attached a video that shows my problem. At first, I thought this was a `nvim-cmp` bug, but I notice this only happens with `tabnine` sources.

### Video

https://user-images.githubusercontent.com/273399/141367402-50c753be-7246-4ceb-8084-d27bcd8ca4b9.mp4

Otherwise, insertion works properly, as long as the insertion doesn't end with a bracket or parenthesis.

**Comment by tzachar** [comment]
Feature, not a bug 😄 
Every time you see a `..` it means its a snippet. The cursor will be placed where the dots indicate, allowing you to fill in the brackets.
You then need to use Tab to navigate to the end of the snippet.

**Comment by mystilleef** [comment]
> Feature, not a bug smile Every time you see a `..` it means its a snippet. The cursor will be placed where the dots indicate, allowing you to fill in the brackets. You then need to use Tab to navigate to the end of the snippet.

Yeah, snippet expansion works pretty well. But in my case, that wasn't a snippet expansion. You can see the completion item `order.id}` does not have a snippet label or placeholder around it. Like you said, snippets are annotated with `..` and `~` to indicate snippet expansion.

**Comment by mystilleef** [comment]
I've added another video showing the problem. You can see when I select the `tabnine` completion item, `cmpapi")`, the cursor does not move to the end of the line. You can also see `cmpapi")` is not a snippet. I'm having this issue when completion items have closing brackets.

### Video

https://user-images.githubusercontent.com/273399/141421806-d8355c6d-f42d-46e5-8242-33b448cd81da.mp4

**Comment by yioneko** [comment]
Same problem here. I think this is an upstream issue. You can revert to hrsh7th/nvim-cmp@8c08bdf7a4e0f218b233765913ff0669358f4a6e for temporary workaround. Upstream repo is under refactoring works on async completion, idk whether it's related to this.

**Comment by tzachar** [comment]
Thanks @yioneko 
@mystilleef Does reverting solves you issue?
If so, can you please open an issue in `nvim-cmp`?
10x.

**Comment by mystilleef** [comment]
> Thanks @yioneko @mystilleef Does reverting solves you issue? If so, can you please open an issue in `nvim-cmp`? 10x.

Yes, reverting solves the issue. I'll open an issue at `nvim-cmp`. Sorry for the trouble, and thanks for your help.

**Comment by hrsh7th** [comment]
I had allowed calling the callback multiple times by requesting from `tzachar`.

I think it cmp-tabnine issue. I guess the tabnine source returns invalid completion item.

**Comment by tzachar** [comment]
@hrsh7th 
This has nothing to do with the callback being called multiple times. This behavior was required for cmp-fuzzy-buffer.

Anyway, if reverting nvim-cmp to an earlier revision fixes the problem, it makes more sense that this is a cmp issue.

@mystilleef 
Can you send me the output of `:messages` after applying the following patch?
This can help debugging the isse:
```diff
diff --git a/lua/cmp_tabnine/source.lua b/lua/cmp_tabnine/source.lua
index a3b38d2..6c2c426 100644
--- a/lua/cmp_tabnine/source.lua
+++ b/lua/cmp_tabnine/source.lua
@@ -312,4 +312,13 @@ Source._on_stdout = function(_, data, _)
        Source.callback = nil;
 end

+function dump(...)
+    local objects = vim.tbl_map(vim.inspect, {...})
+    print(unpack(objects))
+end
+function Source:resolve(completion_item, callback)
+ dump(completion_item)
+  callback(completion_item)
+end
+
 return Source
```

**Comment by tzachar** [comment]
@mystilleef 
I bisected the issue to e1f880b7e0ae3cae251fb6c7901bc856fe546e0b in upstream cmp. Please reopen the issue there.

10x.

**Comment by tzachar** [comment]
Fixed by @hrsh7th in hrsh7th/nvim-cmp@753f5b7c92da0302efffc5ce6780dffe0602bdf3
